### PR TITLE
Adding leverage macro base

### PR DESCRIPTION
**Pull request by wtj2021**
# Leverage Zaps

### CODE REVIEW

File: src/LeverageZapRouterBase.sol [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        bytes calldata _exchangeData
+    ) internal {
+        LeverageMacroOperation memory op;
+
+        op.tokenToTransferIn = address(stETH);
+        op.amountToTransferIn = _marginIncrease;
+        op.operationType = OperationType.AdjustCdpOperation;
+        op.OperationData = abi.encode(_cdp);
+
+        if (_cdp._isDebtIncrease) {
```
Comment (rayeaster) [inline src/LeverageZapRouterBase.sol:90]: does the `_flType` parameter play a role here for `op.swapAfter`? 

For example, if `_flType` is `eBTC`, and with `_cdp._isDebtIncrease=true`, do we still need to swap after FL?

File: src/EbtcLeverageZapRouter.sol [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+    }
+
+    function adjustCdp(
+        bytes32 _cdpId,
+        AdjustCdpParams calldata params,
+        PositionManagerPermit calldata _positionManagerPermit,
+        bytes calldata _exchangeData
+    ) external {
+        _requireMinAdjustment(params._debtChange);
+        _requireMinAdjustment(params._stEthBalanceChange);
```
Comment (rayeaster) [inline src/EbtcLeverageZapRouter.sol:210]: could user `adjustCdp` by only changing debt? which means `_stEthBalanceChange` might be zero?

**Comment by rayeaster** [comment]
quote from @dapp-whisperer 
```
As a general note, we are strongly considering having a “zap router” contract 
that consolidates access to all the zaps through one proxy address. 

This would be non-upgradable, but would make things more modular and easy to manage on the zap development side. 

Only one address outside of the core protocol for users to track and approve. 

It could even call the core protocol (borrowerOperations) directly too.
```

I think it is good idea to provide unified access as the only "zap entrance" to external users and integrators. Diamond-like proxy might fit in here by providing flexibility for new features

File: src/EbtcLeverageZapRouter.sol (hunk 2 above)
Comment (wtj2021) [reply src/EbtcLeverageZapRouter.sol:210]: so the thinking here is that if the user only wants to increase debt without putting any of the proceeds back into the system as additional collateral, then FL is probably not needed and the front-end should route the call to the normal zap router. I am not sure if there's an exception to this?

**Pull request merged**